### ISSUE TITLE
Implement Meta-Learning Framework for Streamflow Forecasting

### ISSUE THREAD

**Issue by sandrohuni** [hydrosolutions/long-term-forecasting#39]
# Meta-Learning Framework Implementation

## Overview
Implement a comprehensive meta-learning system that combines multiple base model predictions through intelligent ensemble weighting and advanced meta-modeling techniques.

## Motivation
Current ensemble approaches use simple averaging of base models (XGBoost, LightGBM, CatBoost). Meta-learning can significantly improve predictions by:
- Learning basin-specific and temporal model performance patterns
- Dynamically weighting models based on historical performance
- Capturing complex interactions between base model predictions
- Providing uncertainty quantification

### Phase 1: Historical Performance-Based Meta-Learning (HistMeta)
Create  class that follows SciRegressor structure:

#### Core Functionality:
- **Performance-based weighting**: Compute historical performance metrics (nRMSE, R2, nMAE) for each base model
- **Basin-temporal-specific weighting**: Different model weights per basin and period (see data_utils.py) based on historical performance.
- **Temporal-specific weighting** If we do not have enough samples for a basin - we weigh by how good the model is usally during that period
- **LOOCV integration**: Proper cross-validation using Leave-One-Year-Out approach

#### Technical Implementation:
Inherits base forecasting model
```python
class HistoricalMetaLearner(BaseForecastModel):
    def __init__(self, base_model_predictions, performance_metrics, ...):
        # Load base model predictions from multiple models
        # Calculate basin-specific and temporal performance weights
        
    def compute_weights(self, metric='rmse'):
        # Compute weights based on historical performance
        # Support basin-specific and temporal weighting
        
    def calibrate_model_and_hindcast(self):
        # Use LOOCV to test weighted ensemble performance
        # Return hindcast predictions with ensemble weights
    def predict_operational(self)
# predict operational discharge 
   def tune_hyperparams(self)
# tune hyperparams - not supported now
```

The base_models should be loaded based on the config: The prediction colums follow this schema  Q_ in col if col != Q_obs, a base model can already be a ensemble for example Snow_GBT has the columns Q_xgb, Q_lgbm, Q_catboost and Q_Snow_GBT -> for the meta model the columns should be restructured to Q_Snow_GBT_xgb, Q_Snow_GBT_lbgm, Q_Snow_GBT_catboost 
 
### File Structure:
```
monthly_forecasting/
├── forecast_models/
│   ├── meta_learners/
│   │   ├── __init__.py
│   │   ├── base_meta_learner.py
│   │   ├── historical_meta_learner.py
│   └── ...
├── scr/
│   ├── meta_utils.py              # Meta-learning utilities
│   └── performance_metrics.py     # Performance calculation -> see dev_tools/evaluation -> use similar approach
└── tests/
    └── Unit /functionality /integration
           └── test_meta_learners.py
```

### Configuration Support:
Add meta-learning configuration to existing config system:
```json
{
  "meta_learning": {
    "enabled": true,
    "type": "historical"
    "base_models": [List of paths],
    "performance_metric": "nrmse",
    "weighting_strategy": "basin_temporal",
  }
}
```

### Integration with Existing Pipeline:
- Extend  to support meta-learning output
- Maintain compatibility with existing model evaluation
- Support operational prediction workflow
- Add meta-learning specific metrics

### Unit Tests:
- Test performance weight calculation
- Test meta-feature creation
- Test LOOCV integration
- Test distributional predictions

### Integration Tests:
- Test full meta-learning pipeline
- Test with different base model combinations
- Test operational prediction mode
- Performance benchmarking vs. simple ensemble

### Validation Approach:
- Use existing LOOCV framework
- Compare against simple averaging baseline
- Analyze performance improvements per basin/season

### Quantitative:
- RMSE, R2, MAE, NSE improvements over baseline
- Uncertainty calibration metrics (if applicable)
- Computational efficiency (training/prediction time)
- Cross-validation consistency

### Qualitative:
- Code maintainability and integration
- Documentation and testing coverage
- Operational usability
- Scientific interpretability

### High Priority (Phase 1): Historical Meta-Learning
- **Timeline**: 2-3 weeks
- **Focus**: Performance-based weighting with LOOCV validation
- **Expected ROI**: High (significant improvement, moderate complexity)

## Dependencies
- Existing SciRegressor implementation
- Base model predictions from XGBoost/LightGBM/CatBoost
- Performance evaluation framework
- LOOCV infrastructure

## Labels
- enhancement
- machine-learning
- meta-learning
- forecasting
- priority-high
